control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,101 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	GatewayClassFinalizer      = "gateway-exists-finalizer.networking.x-k8s.io"
+	GatewayClassControllerName = "hashicorp.com/consul-api-gateway-controller"
+)
+
+// GatewayClassReconciler reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassReconciler struct {
+	ControllerName string
+	Log            logr.Logger
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/finalizers,verbs=update
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if string(gc.Spec.ControllerName) != r.ControllerName {
+		// This GatewayClass is not for this controller.
+		return ctrl.Result{}, nil
+	}
+
+	if !gc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// We have a deletion request. Ensure we are not in use.
+		used, err := r.isGatewayClassInUse(ctx, gc)
+		if err != nil {
+			log.Error(err, "unable to check if GatewayClass is in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			// Requeue after 10 seconds to check again.
+			log.Info("GatewayClass is in use, cannot delete")
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
```

Comment:
same with the implementation of the class config, I think we may want to move towards doing the edge triggering based off of a gateway update v. polling once something is soft deleted, but that can wait for follow-up.

Reply:
Ah, I see. Instead of requeueing, just add Gateway as a triggering type?

Reply:
Yep, but I think that can probably wait for a follow-up in which we handle both this and the gatewayclassconfig controller at the same time. Just moving towards strict edge-triggering for noise and performance v. polling every 10 seconds while something is in use.